묵찌빠게임 [Step2] Kiseok, Morgan

devjoon opened the pull request:
@Hoon94 
안녕하세요 hoon ! 2조 Kiseok, Morgan 입니다. 두번째 스텝 PR보내 드립니다.
이번에도 피드백 주시는 문제에 대해서 잘 수정해보겠습니다 잘 부탁드립니다!

##고민되었던 점
---
- private 을 함수에 사용을 할지말지 고민을 하다가 넣지 않는것으로 결정 하였습니다 넣어주는게 좋을까요?<br> private을 넣어주려다보니 너무 많은 곳에 사용하게 되어 고민이었습니다.

- step2 구현 과정에서 가위바위보와 묵찌빠 두 개의 구조체를 따로 만들어서 사용할지 아니면 비슷한 기능의 함수가 많으니 한 곳에서 비슷한 기능의 함수를 재사용하여 코드의 양을 줄일지 고민되었습니다 지금은 구조체를 따로 만들어서 작업하는 방식으로 진행하였습니다.

carti1108 committed:
feat: 가위바위보 구조체, 묵찌빠 구조체 구현

묵찌빠 구조체에서 printTurn함수 추가 및 사용자와 컴퓨터의 패 만드는 함수 추가
player enum 추가 턴 나타낼 때 사용

devjoon committed:
fix: 오류수정

0~3 사이 숫자가 아닌 다른 숫자나 잘못된 입력을 한 경우 게임을 진행했을때 여러번 출력되던 오류 수정
printMukChiPaSelect 에서 printMukChiPaMenu 로 이름 변경

devjoon committed:
Update README.md

Hoon94 commented on RockPaperScissors/RockPaperScissors/mukChiPaStruct.swift:
```diff
@@ -0,0 +1,110 @@
+//  mukChiPaStruct.swift
+//  RockPaperScissors
+//
+//  Created by 박기석 on 2023/09/07.
```
두 분이서 함께 작업하신 만큼 `enumType` 파일처럼 두 분 모두 이름이 들어가도록 변경해 주시면 좋을 것 같습니다.😊

carti1108 replied:
수정완료했습니다!

Hoon94 commented on RockPaperScissors/RockPaperScissors/main.swift:
```diff
@@ -8,85 +8,96 @@ var userHand: RockPaperScissors = .exit
 var computerHand: RockPaperScissors = .exit
 var rockPaperScissorsResult: gameResult = .draw
 let randomNumberRange = 1...3
+var turn = Player.nobody
```
- 전역 변수를 사용할 수도 있겠지만 `struct`를 사용하신 만큼 [`Initializer`](https://docs.swift.org/swift-book/documentation/the-swift-programming-language/initialization)를 사용해 보시면 어떨까요?😊
- `struct`의 `Initializer`는 어떤 특징이 있을까요?

devjoon replied:
1. 이니셜 라이저를 사용하는 것으로 수정 완료하였습니다
2. struct의 initializer는 class의 initializer와 다르게 memberwiseinitializer 가 있습니다
memberwiseinitializer는 initializer를 사용자가 직접 정의하지 않아도 프로퍼티의 이름을 매개변수로 받도록 해주는 initializer입니다

carti1108 replied:
멤버와이즈이니셜라이저를 사용하려고했으나 이니셜라이저를 사용했다는 것을 직관적으로 보여드리기 위해 일단은 사용자 정의 이니셜라이저를 사용하였습니다!

Hoon94 replied:
`MukChiPa`처럼 `RockPaperScissorsGame`에서도 사용해 보시는 것은 어떨까요?

carti1108 replied:
수정 완료하였습니다!

devjoon committed:
Update README.md

Hoon94 commented on RockPaperScissors/RockPaperScissors/enumType.swift:
```diff
@@ -11,6 +11,23 @@ enum RockPaperScissors {
+    }
+}
+
 enum gameResult {
```
타입명에 대한 컨벤션을 지켜주세요.🙏🏼

- [Api design guidelines](https://www.swift.org/documentation/api-design-guidelines/#conventions)

carti1108 replied:
GameResult로 변경완료하였습니다!

Hoon94 commented on RockPaperScissors/RockPaperScissors.xcodeproj/project.pbxproj:
```diff
@@ -7,6 +7,7 @@
 
 /* Begin PBXBuildFile section */
+		4F916F022AA99C2700CEF695 /* mukChiPaStruct.swift in Sources */ = {isa = PBXBuildFile; fileRef = 4F916F012AA99C2700CEF695 /* mukChiPaStruct.swift */; };
 		927DDAE32AA7068D00AACA37 /* enumType.swift in Sources */ = {isa = PBXBuildFile; fileRef = 927DDAE22AA7068D00AACA37 /* enumType.swift */; };
```
파일명에 대한 컨벤션을 지켜주세요.🙏🏼

carti1108 replied:
MukChiPaStruct, EnumType으로 변경완료하였습니다!

Hoon94 commented on RockPaperScissors/RockPaperScissors/main.swift:
```diff
@@ -8,85 +8,96 @@ var userHand: RockPaperScissors = .exit
-        try convertToUserHand(user: userInput)
-    } catch {
-        print("잘못된 입력입니다. 다시 시도해주세요")
+struct RockPaperScissorsGame {
```
`struct` 타입을 사용하셨네요.👍
`struct`를 선택하신 이유는 무엇일까요?

carti1108 replied:
struct는 데이터들을 용도에 맞게 묶어 캡슐화를 할때 용이하다고 생각했습니다.
따라서 가위바위보와 묵찌빠를 두 개로 나눠 캡슐화를 하려고 했습니다.

Hoon94 replied:
[`struct`와 `class`](https://docs.swift.org/swift-book/documentation/the-swift-programming-language/classesandstructures/)는 어떤 차이가 있을까요?🤔

- 참고 자료 : [Choosing Between Structures and Classes](https://developer.apple.com/documentation/swift/choosing-between-structures-and-classes)

carti1108 replied:
1. struct는 값타입 class는 참조타입입니다. 값타입이란 값을 복사하여 사용하는 것이고, 참조타입이란 값을 참조하여 사용하는 것입니다. 따라서 값타입에서의 값 변경 시 원본 값에는 영향이 없지만 참조타입에서 값 변경 시에는 원본값에 영향을 줍니다.
2. struct는 memberwiseinitializer가 있지만 class는 없습니다.
3. struct는 상속이 불가능하지만 class는 상속이 가능합니다.
위와 같은 이유들로 인해 상속할 필요가 없고 값을 복사해서 사용하고 싶은 경우에는 struct를 사용하는 것이 좋다고 생각합니다. 이 경우가 아니라면 메모리 낭비를 줄여주는 class를 사용하는 것이 더 좋다고 생각합니다.

Hoon94 commented on RockPaperScissors/RockPaperScissors/mukChiPaStruct.swift:
```diff
@@ -0,0 +1,110 @@
+//  Created by 박기석 on 2023/09/07.
+//
+
+import Foundation
```
- `Foundation`의 역할은 무엇인가요?🤔
- 지금 코드에서 필요한 라이브러리일까요?
- `enumType.swift` 파일에도 남아있었네요.🥲

carti1108 replied:
Foundation은 데이터 저장 및 지속성, 텍스트 처리, 날짜 및 시간 계산, 정렬 및 필터링, 네트워킹을 포함하여 앱 및 프레임워크에 대한 기본 기능 계층을 제공합니다.
전에 localizedError 사용하려고 했었던 부분이었는데 빼는 과정에서 잊고있었던것 같습니다.
제거 완료하였습니다!

Hoon94 commented on RockPaperScissors/RockPaperScissors/mukChiPaStruct.swift:
```diff
@@ -0,0 +1,110 @@
+    }
+    
+    func printMukChiPaMenu(winner: Player) {
+        print("[\(winner.player) 턴] 묵(1), 찌(2), 빠(3)! <종료 : 0>", terminator: " : ")
```
[`String Interpolation`](https://docs.swift.org/swift-book/documentation/the-swift-programming-language/stringsandcharacters/#String-Interpolation)을 사용하셨네요.😊 가독성이 좋아 보입니다.👍

Hoon94 commented on RockPaperScissors/RockPaperScissors/mukChiPaStruct.swift:
```diff
@@ -0,0 +1,110 @@
+        }
+    }
+    
+    func printMukChiPaMenu(winner: Player) {
```
기존에 사용하던 메서드 배치 컨벤션과는 달라보이네요. 이렇게 배치하신 다른 이유가 있을까요?🤣

devjoon replied:
확인을 하지 못한 부분 인것 같습니다 꼼꼼한 피드백 감사드립니다. 배치 컨벤션 수정 완료하였습니다

Hoon94 commented on RockPaperScissors/RockPaperScissors/mukChiPaStruct.swift:
```diff
@@ -0,0 +1,110 @@
+    }
+    
+    func generateComputerHand() {
+        let randomComputerHand = Int.random(in: 1...3)
```
`RockPaperScissorsGame`에서 처럼 매직 리터럴을 줄여보시면 좋을것 같습니다.🤗

devjoon replied:
이전에 사용했던 randomNumberRange를 이용하여 매직 리터럴 처리 완료하였습니다. 감사합니다

Hoon94 reviewed (commented):
안녕하세요. Kiseok, Morgan👋
STEP 2 진행까지 고생 많으셨습니다.👍
마지막 리뷰 답변까지 파이팅입니다.🔥

## **칭찬 드리고 싶은 부분**

- `git branch`를 사용하여 각 STEP 별로 나누어서 진행하신 점👍
- 각각의 함수에 대해 기능을 분리하려고 노력하신 점👍
- 타입을 사용하여 코드를 분리하신 점👍
- 각각의 커밋을 자세히 작성하신 점👍

## 고민되었던 점

> Q: private 을 함수에 사용을 할지말지 고민을 하다가 넣지 않는것으로 결정 하였습니다 넣어주는게 좋을까요? 
private을 넣어주려다보니 너무 많은 곳에 사용하게 되어 고민이었습니다.

A: `private`의 역할은 무엇일까요? ``private``을 사용했을 때의 장점에 대해 설명해 주실 수 있을까요?

> Q: step2 구현 과정에서 가위바위보와 묵찌빠 두 개의 구조체를 따로 만들어서 사용할지 아니면 비슷한 기능의 함수가 많으니 한 곳에서 비슷한 기능의 함수를 재사용하여 코드의 양을 줄일지 고민되었습니다 지금은 구조체를 따로 만들어서 작업하는 방식으로 진행하였습니다.

A: 코드의 중복을 줄이는 일은 가독성을 높이고 일반화를 진행하여 코드의 재사용성을 늘리는 등 다양한 장점이 있다고 생각합니다. 말씀해 주신 것처럼 비슷한 기능을 묶어서 관리한다면 더 좋은 사용성을 가질 수 있을 것 같습니다.

## Tip

- 커밋 내용에 대한 메시지를 상세히 달아주셔서 충분히 각 커밋에서 어떤 작업을 수행하셨는지 이해할 수 있었습니다. 내용을 기준으로 본다면 충분히 어떤 커밋에서 무엇을 했는지를 파악할 수 있다고 생각합니다. 하지만 다음에는 상세 설명을 적기보다는 각각을 하나의 커밋으로 나누어서 진행해 보면 좋을 것 같습니다. 커밋을 작은 단위로 세분화하는 작업은 커밋 내역을 쉽게 받아들일 수 있다는 장점 외에도 이전 버전으로 돌아가고 싶을 때 최소한으로만 이동할 수 있는 안전장치를 만들어 둔다고 볼 수 있습니다. 다음 프로젝트에서는 한 번 시도해 보시는 걸 추천드립니다.😊

Hoon94 commented on RockPaperScissors/RockPaperScissors/main.swift:
```diff
@@ -8,85 +8,96 @@ var userHand: RockPaperScissors = .exit
+    func generateResult() {
+        if computerHand == userHand {
+            rockPaperScissorsResult = .draw
+            print(rockPaperScissorsResult.result)
```
저도 네이밍은 항상 너무 어려운 것 같습니다.😭
`rockPaperScissorsResult.result`에서 `Result.result`가 연속적으로 사용되고 있습니다.
코드를 사용할 때 전체적인 코드의 문맥을 생각해 보고 각각의 네이밍을 작성하는 것도 좋은 방법일 것 같습니다.🤩
어떻게 네이밍을 바꿀 수 있을까요?

Hoon94 commented on RockPaperScissors/RockPaperScissors/main.swift:
```diff
@@ -8,85 +8,96 @@ var userHand: RockPaperScissors = .exit
+    func generateResult() {
+        if computerHand == userHand {
+            rockPaperScissorsResult = .draw
+            print(rockPaperScissorsResult.result)
```
저도 네이밍은 항상 너무 어려운 것 같습니다.😭
`rockPaperScissorsResult.result`에서 `Result.result`가 연속적으로 사용되고 있습니다.
코드를 사용할 때 전체적인 코드의 문맥을 생각해 보고 각각의 네이밍을 작성하는 것도 좋은 방법일 것 같습니다.🤩
`Player.player`도 같이 생각해 보시면 좋겠습니다.
어떻게 네이밍을 바꿀 수 있을까요?

carti1108 replied:
result, player에서 resultDescription, playerDescription으로 변경하였습니다!

Hoon94 commented on RockPaperScissors/RockPaperScissors/mukChiPaStruct.swift:
```diff
@@ -0,0 +1,110 @@
+            rockPaperScissorsResult = .exit
+            print(rockPaperScissorsResult.result)
+        } else {
+            printTurn(player: Player.computer)
```
열거형을 사용하는 과정에서 `RockPaperScissors` 타입에서는 `.rock`과 같이 사용하시고 `Player` 타입에서는 `Player.computer`와 같이 사용하셨습니다.
하나의 컨벤션으로 통일해 주시면 더 좋을 것 같습니다.🙏🏼

carti1108 replied:
컨벤션 통일 완료하였습니다!

carti1108 committed:
파일,타입명 변경
enumType,mukChiPaStruct파일을 EnumType,MukChiPaStruct로 변경
gameResult 열거형을 GameResult로 변경
Player,GameResult의 연산 프로퍼티명을 playerDescription, resultDescription으로 변경

devjoon committed:
feat: init 추가

carti1108 committed:
MukChiPa 구조체에 turn 변수 만들어주고 initializer 사용

devjoon committed:
refactor: 함수 배치수정 , 매직리터럴 제거

carti1108 committed:
refactor: 컨벤션 수정

generateResult메서드 부분 컨벤션 통일

carti1108 committed:
refactor: MukChiPa 구조체의 init부분 수정

carti1108 commented:
private 키워드는 가장 한정적인 범위로 그 기능을 정의한 범위에서만 사용할 수 있습니다.
private을 사용하면 남이 접근해서는 안되는 정보를 은닉화 할 수 있는 장점이 있습니다.

devjoon committed:
fix: 가위바위보에서 졌을때 묵찌빠가 사용자턴으로 실행되던 오류 수정

devjoon committed:
refactor: 함수배치변경

devjoon committed:
refactor: enum 프로퍼티명 변경

Player, GameResult의 description 부분 네이밍 변경

carti1108 committed:
refactor: 전역변수를 struct의 프로퍼티로 변경

전역변수였던 userHand, computerHand, rockPaperScissorsResult를 구조체의 프로퍼티로 변경 후 이니셜라이저 사용

devjoon committed:
Update README.md

devjoon committed:
Update README.md

devjoon committed:
Update README.md

devjoon committed:
Update README.md

devjoon committed:
Update README.md

devjoon committed:
Update README.md

devjoon committed:
Update README.md

Hoon94 commented:
한 주 동안 고생 많으셨습니다. 남은 캠프 생활도 파이팅입니다.🔥
